Update docs (#28)
SovereignCloudStack/k8s-harbor: docs/harbor_backup_restore.md

# Harbor - backup and restore

This page aims at providing a step-by-step guide for backup and restore [Harbor](https://goharbor.io/)
container registry using [Velero](https://velero.io/) tool.
It extends the official [Harbor backup-restore docs page](https://goharbor.io/docs/main/administration/backup-restore/)
with up-to-date commands, explanations, and an extensive prerequisites section. This
guide references and uses Velero in [v1.10.2](https://github.com/vmware-tanzu/velero/releases/tag/v1.10.2)

> as this is the latest stable version at the time of writing this guide.

Reviewer: > this is the latest stable version at the time of writing this guide

I like this disclaimer.

In the general context of maintenance: Is it the goal to provide docs which are "future-proof" or is it okay to serve as a reference of "this worked at the time of writing"? @maxwolfs ?

(When thinking about future-proofing documentation, I'd land at continuously "unittesting" documentation and using e.g. renovate to keep it up to date. As renovate/CI-testing has barely landed in the more code-centric contexts in SCS, this of course would be a topic for some time in the future - not for now anyway.)

Not a blocking problem for this PR anyway.

Author: The idea of future-proofing documentation with a continuous "unittesting" pipeline is interesting from my point of view.   IMO, it may require a serious effort for writing CI pipelines that should cover possibly various scenarios that could be mentioned in the docs.
   
@maxwolfs, it would be great to know your opinion here.

Reviewer 2: Thanks for bringing this up. I think this is a good topic to discuss further within the next SIG Documentation on the 15th of May at 11:05 at SCS-Tech. 

In the [2023-02-20](https://github.com/SovereignCloudStack/minutes/blob/main/sig-documentation/20230220.md#version-current-docs-for-upcoming-release) date we discussed versioning and came to the conclusion that we only want to serve _latest_ documentation.

From that point of view, I expect that the docs will always work and if they do not because of anything stops working, code & docs have to be updated. Then `at the time of writing this guide` is somehow obsolete from my point of view because it applies for the whole docs, but makes of course sense in combination with mentioning specific versions. 

I don't see this as a blocker too. Let's merge :)

Author: @joshmue could we resolve this one?

Reviewer: Sure!

(even if the PR could be merged either way, of course)

It provides guidance and commands that readers are encouraged to try out by themselves
on Harbor deployment as described in the next sections. It does not aim at providing an
exhaustive list of commands nor all the possible ways how to use them.

The guide covers two strategies to save Harbor data:
- Backup: a regular backup created by the [restic](https://restic.net/) integration in Velero as described in the related [docs](https://velero.io/docs/main/file-system-backup/)
- Snapshot: a point-in-time snapshot be the Container Storage Interface (CSI) snapshot support in Velero as described in the related [docs](https://velero.io/docs/main/csi/)

> Before you choose the right strategy for your Harbor deployment back up make sure that you understand

Reviewer: ```suggestion
Before you choose the right strategy for your Harbor deployment backup, make sure that you understand
```

[differences](https://www.terra-master.com/global/snapshot) between the backup and snapshot.
In general, for long-term protection of Harbor data, you may use backup and for
temporary protection of data (e.g. before Harbor upgrade) you may use snapshot.

Note that this guide is not limited to Harbor deployments that utilize SCS environments,
but it is required to have a set of tools and services (e.g. Kubernetes CSI plugin with volume snapshot
support, S3 compatible object store for backups) for successful backup and restore procedure
(see the [prerequisites](#prerequisites-) section). These tools and services come out of
the box when the SCS infrastructure and KaaS is used for Harbor deployment, hence it is
convenient to use them. 

## Prerequisites 

### Kubernetes cluster

If you want to use `snapshot` to back up Harbor data ensure the following:
- Your cluster is Kubernetes version 1.20 or greater
- Your cluster is running a [CSI driver](https://kubernetes-csi.github.io/docs/drivers.html)
  capable of support volume snapshots at the [v1 API level](https://kubernetes.io/blog/2020/12/10/kubernetes-1.20-volume-snapshot-moves-to-ga/).

>   To enable creating volume snapshots the [snapshot-controller](https://kubernetes-csi.github.io/docs/snapshot-controller.html)

Reviewer: ```suggestion
  To enable creating volume snapshots, the [snapshot-controller](https://kubernetes-csi.github.io/docs/snapshot-controller.html)
```

  and its CRDs should be deployed in the Kubernetes cluster as well. The snapshot-controller
  is independent of any CSI Driver. These prerequisites come out of the box with the SCS KaaS solution.

If you want to create Harbor `backup` ensure the following:  
- Your cluster is Kubernetes version 1.16 or greater

If your cluster meets the above, export its kubeconfig path in env. variable `KUBECONFIG`:
```bash
export KUBECONFIG=/path/to/kubeconfig
```

### S3 bucket and EC2 credentials

This guide assumes that the public cloud's object store with S3-compatible API is available as
the storage backend for Velero. In this guide, we are using OpenStack Swift which 
offers S3-compatible API. Let's create an S3 bucket on Swift object storage and EC2 credentials
that will be later used by Velero.

You should have access to your OpenStack project, and the OpenStack RC file that contains access 
values. Set the environment variables by sourcing the OpenStack RC file:

```bash
source <project>-openrc.sh
```

Swift object store service does not support application credentials authentication
to access S3 API. To authenticate in S3 API, you should generate and use the EC2 credentials

> mechanism. You can generate EC2 credentials as follows:

Reviewer: To elaborate on that: AFAIU, this means that access is directly bound to the user doing these steps, which has a load of "natural" disadvantages.

Let's document this and/or create an issue with that topic: Making application credentials work or (even better) making machine/workload identities work.

Author: I agree, that it is a good idea to mention such disadvantages of  EC2 credentials. See f5896e66a94456185911801b6a66b1ac790c73cc where I warn users and also suggest a potential workaround.

command and store them in the `~/.aws/credentials` file as follows:

```bash
mkdir ~/.aws
cat >~/.aws/credentials <<EOF
[default]
aws_access_key_id = <aws_access_key_id>
aws_secret_access_key = <aws_secret_access_key>
EOF
```

This credential file is then used as an access and secret source for AWS CLI tool and also
as a source for Velero. If your environment does not have AWS CLI installed, install it as follows:

```bash
pip3 install awscli awscli-plugin-endpoint
```

Finally, create a new bucket. Note that the following command contains `endpoint-url`
argument that points AWS CLI to the GX-SCS OpenStack Swift object store API.

```bash
aws --endpoint-url https://api.gx-scs.sovereignit.cloud:8080 s3 mb s3://velero-backup
```

### Velero client

In this guide, we are using [Velero](https://velero.io/) to back up and restore a Harbor instance. 
Velero is an open source tool to safely back up and restore, perform disaster recovery,
and migrate Kubernetes cluster resources.

Go through the [official docs](https://velero.io/docs/main/basic-install/) and 
install the Velero client on your desired environment. If your environment is Linux distribution

> you can use the following steps and install the Velero client from the source:

Reviewer: ```suggestion
you can use the following steps and install the Velero client from the Github release binaries:
```

  - Install Velero:
  ```bash
  velero install \
    --provider aws \
    --plugins velero/velero-plugin-for-aws:v1.6.1 \
    --bucket velero-backup \
    --secret-file ~/.aws/credentials \
    --use-volume-snapshots=false \
    --uploader-type=restic \
    --use-node-agent \
    --backup-location-config region=RegionOne,s3ForcePathStyle="true",s3Url=https://api.gx-scs.sovereignit.cloud:8080
  ```

##  Backup and restore

Note that the following backup steps mainly point to actions from an official [Backup And Restore Harbor With Velero](https://goharbor.io/docs/main/administration/backup-restore) tutorial.
In this guide, find the added value from additional explanations/hints and up-to-date commands.

Harbor, by design, consists of multiple (micro)services that could store their data
variously, based on the Harbor configuration. See the [Harbor persistence](harbor_persistence.md)
docs for further information regarding the Harbor persistence layer. The following steps cover
cases when Harbor persistence is enabled and the "internal" databases (PostgreSQL and Redis)
are used.

Note that backups of external databases (PostgreSQL and Redis) are not supported by the 

> official tutorial and are out of the scope of this guide as well.

Reviewer: This left my wondering if backing up postgres and redis is required at all. Does it only hold barely important metadata or is it mandatory for restoring?

Please consider clarifying this explicitly on the guide.

Author: Explained in more detail within fixup f5896e66a94456185911801b6a66b1ac790c73cc

exhaustive list of commands nor all the possible ways how to use them.

The guide covers two strategies to save Harbor data:
- Backup: a regular backup created by the [restic](https://restic.net/) integration in Velero as described in the related [docs](https://velero.io/docs/main/file-system-backup/)
- Snapshot: a point-in-time snapshot be the Container Storage Interface (CSI) snapshot support in Velero as described in the related [docs](https://velero.io/docs/main/csi/)

Before you choose the right strategy for your Harbor deployment backup, make sure that you understand
[differences](https://www.terra-master.com/global/snapshot) between the backup and snapshot.
In general, for long-term protection of Harbor data, you may use backup and for
temporary protection of data (e.g. before Harbor upgrade) you may use snapshot.

Note that this guide is not limited to Harbor deployments that utilize SCS environments,
but it is required to have a set of tools and services (e.g. Kubernetes CSI plugin with volume snapshot
support, S3 compatible object store for backups) for successful backup and restore procedure

> (see the [prerequisites](#prerequisites-) section). These tools and services come out of

Reviewer 3: ```suggestion
(see the [prerequisites](#prerequisites) section). These tools and services come out of
```

> the box when the SCS infrastructure and KaaS is used for Harbor deployment, hence it is

Reviewer 3: ```suggestion
the box when the SCS infrastructure and KaaS are used for Harbor deployment, hence it is
```

convenient to use them. 

> ## Prerequisites 

Reviewer 3: ```suggestion
## Prerequisites
```

command and store them in the `~/.aws/credentials` file as follows:

```bash
mkdir ~/.aws
cat >~/.aws/credentials <<EOF
[default]
aws_access_key_id = <aws_access_key_id>
aws_secret_access_key = <aws_secret_access_key>
EOF
```

This credential file is then used as an access and secret source for AWS CLI tool and also
as a source for Velero. If your environment does not have AWS CLI installed, install it as follows:

```bash
pip3 install awscli awscli-plugin-endpoint
```

Finally, create a new bucket. Note that the following command contains `endpoint-url`
argument that points AWS CLI to the GX-SCS OpenStack Swift object store API.

```bash
aws --endpoint-url https://api.gx-scs.sovereignit.cloud:8080 s3 mb s3://velero-backup
```

### Velero client

In this guide, we are using [Velero](https://velero.io/) to back up and restore a Harbor instance. 
Velero is an open source tool to safely back up and restore, perform disaster recovery,
and migrate Kubernetes cluster resources.

Go through the [official docs](https://velero.io/docs/main/basic-install/) and 
install the Velero client on your desired environment. If your environment is Linux distribution

> you can use the following steps and install the Velero client from the Github release binaries:

Reviewer 3: ```suggestion
you can use the following steps and install the Velero client from the GitHub release binaries:
```
